test/test_ekfac.py

```diff
@@ -0,0 +1,1094 @@
+        shuffle: Whether to shuffle the parameters before computing the EKFAC matrix.
+        exclude: Which parameters to exclude. Can be ``'weight'``, ``'bias'``,
+            or ``None``.
+        separate_weight_and_bias: Whether to treat weight and bias as separate blocks in
+            the EKFAC matrix.
+    """
+    model, loss_func, params, data, batch_size_fn = kfac_exact_case
+    params = maybe_exclude_or_shuffle_parameters(params, model, exclude, shuffle)
+
+    ggn = block_diagonal(
+        GGNLinearOperator,
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+    ekfac = EKFACLinearOperator(
+        model,
+        loss_func,
+        params,
+        data,
+        batch_size_fn=batch_size_fn,
+        fisher_type=FisherType.TYPE2,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+    ekfac_mat = ekfac.to_scipy() @ eye(ekfac.shape[1])
+
+    report_nonclose(ggn, ekfac_mat, atol=1e-6)
+
+    # Check that input covariances were not computed
+    if exclude == "weight":
+        assert len(ekfac._input_covariances_eigenvectors) == 0
+
+
+@mark.parametrize("setting", [KFACType.EXPAND, KFACType.REDUCE])
+@mark.parametrize(
+    "separate_weight_and_bias", [True, False], ids=["separate_bias", "joint_bias"]
+)
+@mark.parametrize(
+    "exclude", [None, "weight", "bias"], ids=["all", "no_weights", "no_biases"]
+)
+@mark.parametrize("shuffle", [False, True], ids=["", "shuffled"])
+def test_ekfac_type2_weight_sharing(
+    kfac_weight_sharing_exact_case: Tuple[
+        Union[WeightShareModel, Conv2dModel],
+        MSELoss,
+        List[Parameter],
+        Dict[str, Iterable[Tuple[Tensor, Tensor]]],
+    ],
+    setting: str,
+    shuffle: bool,
+    exclude: str,
+    separate_weight_and_bias: bool,
+):
+    """Test EKFAC for linear weight-sharing layers against the exact GGN.
+
+    Args:
+        kfac_weight_sharing_exact_case: A fixture that returns a model, loss function, list of
+            parameters, and data.
+        setting: The weight-sharing setting to use. Can be ``KFACType.EXPAND`` or
+            ``KFACType.REDUCE``.
+        shuffle: Whether to shuffle the parameters before computing the EKFAC matrix.
+        exclude: Which parameters to exclude. Can be ``'weight'``, ``'bias'``,
+            or ``None``.
+        separate_weight_and_bias: Whether to treat weight and bias as separate blocks in
+            the EKFAC matrix.
+    """
+    model, loss_func, params, data, batch_size_fn = kfac_weight_sharing_exact_case
+    # The model outputs have to be flattened assuming only the first dimension is the
+    # batch dimension since EKFAC only supports 2d outputs.
+    model.setting = "expand-flatten" if "expand" in setting else setting
+    if isinstance(model, Conv2dModel):
+        # parameters are only initialized after the setting property is set
```

Comment:
Not sure I understand how the comment connects to the next line of code here.

Reply:
I clarified the comment.